### PR TITLE
fix the problem of sigmoid gradient generating NaN

### LINKED ISSUE #1139: Small negative values cause the gradient of `sigmoid` to become `NaN`
Here is the code:
```rust
use burn::{
    backend::{Autodiff, NdArray},
    tensor::{activation, Data, Tensor},
};

fn main() {
    let data = Data::<f32, 1>::from([-90.0]);

    let device = Default::default();
    let tensor_1 = Tensor::<Autodiff<NdArray>, 1>::from_data(data, &device).require_grad();

    let tensor_2 = activation::sigmoid(tensor_1.clone());
    let grads = tensor_2.backward();

    let grad_1 = tensor_1.grad(&grads).unwrap();
    println!("{}", grad_1);
}
```
The result is `NaN`.

### PR DESCRIPTION
**Pull request by wcshds**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

fix #1139 

### Changes

Use sigmoid's derivative formula directly to avoid differentiating `log` and `exp` in autodiff.

### CODE REVIEW

File: burn-tensor/src/tensor/ops/tensor.rs [hunk 1]
```diff
@@ -926,6 +926,22 @@ pub trait TensorOps<B: Backend> {
...
+    ///
+    /// # Returns
+    ///
+    /// A tensor with the same shape as `tensor` with sigmoid values.
+    fn sigmoid<const D: usize>(tensor: FloatTensor<B, D>) -> FloatTensor<B, D> {
+        B::exp(B::neg(B::log(B::add_scalar(
+            B::exp(B::neg(tensor)),
+            1.0_f32.elem(),
+        ))))
+    }
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/ops/tensor.rs:943]: We can add this function in `burn-tensor/src/tensor/ops/activation.rs` instead.

File: burn-tensor/src/tensor/activation/base.rs [hunk 2]
```diff
@@ -78,7 +78,14 @@ pub fn log_softmax<const D: usize, B: Backend>(tensor: Tensor<B, D>, dim: usize)
...
 pub fn sigmoid<const D: usize, B: Backend>(tensor: Tensor<B, D>) -> Tensor<B, D> {
-    log_sigmoid(tensor).exp()
+    match B::FloatElem::precision() {
+        Precision::Half => {
+            let tensor_full = tensor.to_full_precision();
+            let tensor_tmp = tensor_full.sigmoid();
+            Tensor::from_full_precision(tensor_tmp)
+        }
+        _ => tensor.sigmoid(),
+    }
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/activation/base.rs:88]: I don't think we need full precision here, as it is now declared as a method in the backend. The backend implementations can choose to use full precision regardless of the circumstances. Perhaps we can consider incorporating full precision into the default implementation.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]

File: burn-tensor/src/tensor/ops/activation.rs [hunk 3]
```diff
@@ -102,4 +103,43 @@ pub trait ActivationOps<B: Backend> {
...
+    fn sigmoid<const D: usize>(tensor: FloatTensor<B, D>) -> FloatTensor<B, D> {
+        let tensor_full = B::to_full_precision(&tensor);
+        let tensor_tmp = B::FullPrecisionBackend::exp(B::FullPrecisionBackend::neg(
+            B::FullPrecisionBackend::log(B::FullPrecisionBackend::add_scalar(
+                B::FullPrecisionBackend::exp(B::FullPrecisionBackend::neg(tensor_full)),
+                1.0.elem(),
+            )),
+        ));
+
+        B::from_full_precision(tensor_tmp)
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/ops/activation.rs:125]: Not sure @louisfd if there is a more numerically stable implementation possible here.

**Review by louisfd — CHANGES_REQUESTED** [review]
Hi, can you check my comment about the naming of the argument to backward, afterwards it will be ready

File: burn-tensor/src/tensor/ops/activation.rs [hunk 4]
```diff
@@ -102,4 +103,43 @@ pub trait ActivationOps<B: Backend> {
...
+    /// # Arguments
+    ///
+    /// * `x` - The tensor.
+    /// * `grad` - The gradient.
+    ///
+    /// # Returns
+    ///
+    /// The output tensor.
+    fn sigmoid_backward<const D: usize>(
+        x: FloatTensor<B, D>,
```
[the remark this answers is not in the document]
Comment (wcshds) [reply burn-tensor/src/tensor/ops/activation.rs]: Done.

**Review by louisfd — APPROVED** [review]
Thanks a lot

**Pull request merged**